### PR TITLE
[BB-3670] Include only new instances in send_trial_instances_report task

### PR DESCRIPTION
**Pull request by spokerman12**
**Description**

We're fulfilling the following user story:

> "As the person converting Ocim trial users, I need an autogenerated monthly report with a list of trial users for each month. "

**Supporting information**

- [Jira ticket](https://tasks.opencraft.com/browse/BB-3670)

**Testing instructions**

1. Set up a local OCIM environment using this branch
2. Create instances and appservers
3. Edit their created dates with any method available
4. Run `send_trial_instances_report` on `reports/tasks.py` or
5. Run `./manage.py test reports.tests.ReportTaskTestCase` inside vagrant VM

**Snapshots:**

None

**Further considerations:**

- Initially we enabled a flag for the command, but we updated the feature to reflect current requirements as discussed in the review comments.

Reviewer:
[ ] @lgp171188

### CODE REVIEW

File: instance/management/commands/instance_statistics_csv.py
Comment (lgp171188) [inline instance/management/commands/instance_statistics_csv.py:101]: @spokerman12, the management command should be left as-is and the code [here](https://github.com/open-craft/opencraft/blob/master/reports/tasks.py) should be updated to do the necessary filtering.

Comment (spokerman12) [reply instance/management/commands/instance_statistics_csv.py:101]: I don't understand, on this [comment](https://tasks.opencraft.com/browse/BB-3670?focusedCommentId=188674&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-188674) you said 

> [implement] a flag to the existing command that takes in a date range and filters the instances down to that

- Are domains always required, then?
- Should I remove the --created flag?
- Should I create another db_periodic_task or edit the current send_trial_instances_report? The name is confusing as it suggests I should edit it.

Comment (lgp171188) [reply instance/management/commands/instance_statistics_csv.py:101]: @spokerman12, yes, I said what you have quoted above. But I feel that it is better to leave the command as-is and generic enough and update the periodic task instead to filter and pass only the instances created in the most recent completed month. Does this make sense or do you have any questions?

Comment (spokerman12) [reply instance/management/commands/instance_statistics_csv.py]: Makes sense! I implemented the changes suggested plus a test.

File: reports/tasks.py
Comment (lgp171188) [inline reports/tasks.py:52]: @spokerman12, nit: the new instances are still trial instances :)

File: reports/tasks.py
Comment (lgp171188) [inline reports/tasks.py:50]: @spokerman12, is there a reason you changed the formatting of this line? The old one met the code quality checks and didn't cause any issues before.

And if there is some reason to change it, you should format it according to our coding standards, like what you had to do for the django-password-validators PR.

File: reports/tasks.py
Comment (lgp171188) [inline reports/tasks.py:59]: @spokerman12, similar concerns as https://github.com/open-craft/opencraft/pull/711/files#r579396180

File: reports/tasks.py
Comment (lgp171188) [inline reports/tasks.py:80]: @spokerman12, since this is just a time range, how much effort would it require to make this configurable instead of hard-coding it to the last completed month?

File: reports/tasks.py
Comment (lgp171188) [inline reports/tasks.py:82]: @spokerman12, nit: closing parenthesis formatting.

File: reports/tasks.py
Comment (lgp171188) [inline reports/tasks.py:103]: @spokerman12, similar concerns as https://github.com/open-craft/opencraft/pull/711/files#r579396180

File: reports/tasks.py
Comment (lgp171188) [inline reports/tasks.py:84]: @spokerman12, I think `instance.domain` will always contain the right value and so this `or` condition may not be needed.

File: reports/tests.py
Comment (lgp171188) [inline reports/tests.py:581]: @spokerman12, the repeated code below for creating multiple users can be refactored into a separate function or a factory function, no?

File: reports/tests.py
Comment (lgp171188) [inline reports/tests.py:589]: @spokerman12, I don't think `PENDING` is the correct state. Once an appserver is created, I think the status changes to "approved".

File: reports/tasks.py
Comment (spokerman12) [reply reports/tasks.py:52]: Understood. Changed like so.

File: reports/tasks.py
Comment (spokerman12) [reply reports/tasks.py:50]: I ran this through autopep8, but it's true that while we should adhere to PEP8 per the handbook, the 79 character per line limit is changed to 120.

File: reports/tasks.py
Comment (spokerman12) [reply reports/tasks.py:59]: I ran this through autopep8, but it's true that while we should adhere to PEP8 per the handbook, the 79 character per line limit is changed to 120.

File: reports/tasks.py
Comment (spokerman12) [reply reports/tasks.py:80]: It'd be easy, just adding more arguments to send_trial_instances_report.
But right now I don't think it's convenient to increase the scope as we still have tasks left; I'd be happy to do it on the next sprint, though.

File: reports/tasks.py
Comment (spokerman12) [reply reports/tasks.py:82]: Changed according to previous suggestions.

File: reports/tasks.py
Comment (spokerman12) [reply reports/tasks.py:84]: This part of the logic was already implemented, so I wouldn't want to rework it unnecessarily; plus, I don't think the command is tested thoroughly. As suggested before, I'd be more than happy to work on this on a different task.

File: reports/tasks.py
Comment (spokerman12) [reply reports/tasks.py:103]: Corrected as suggested.

File: reports/tests.py
Comment (spokerman12) [reply reports/tests.py:581]: Because it's implied that there's an ordering  user1 < user2 < user3, instance1 < instance2 < instance3 ... etc, because we're adding a singular feature I prefer the test to be explicit (or "don't make me think," as suggested in the handbook).

I'd add the factory function if we were adding more tests that could make use of it, but in this case I don't think it's necessary.

File: reports/tests.py
Comment (spokerman12) [reply reports/tests.py:589]: The command only wants that the referred BetaTestApplication is not null, but I'll change it to Approved to reflect real scenarios.

File: reports/tasks.py
Comment (lgp171188) [inline reports/tasks.py:79]: @spokerman12, this has the off-by-one error as mentioned [here](https://docs.djangoproject.com/en/3.1/ref/models/querysets/#range). Can you fix it and also write a test case for this scenario?

File: reports/tests.py
Comment (lgp171188) [inline reports/tests.py:587]: @spokerman12, this test doesn't test the corner cases like the one mentioned [here](https://github.com/open-craft/opencraft/pull/711/files#r580161566).

File: reports/tasks.py
Comment (spokerman12) [reply reports/tasks.py]: Thank you for picking up this case. I've modified the filter and added the unit test

File: reports/tests.py
Comment (spokerman12) [reply reports/tests.py:587]: I added the test case as stated in previous comment

File: reports/tests.py
Comment (lgp171188) [inline reports/tests.py:664]: @spokerman12, nit: this should be removed.

Comment (spokerman12) [reply reports/tests.py:664]: My bad!

File: reports/tests.py
Comment (lgp171188) [inline reports/tests.py:553]: @spokerman12, nit: formatting of closing parenthesis doesn't match the coding standards that we follow.

Comment (spokerman12) [reply reports/tests.py]: Fixed plus another occurrence.

File: reports/tests.py
Comment (lgp171188) [inline reports/tests.py:599]: @spokerman12, the key doesn't matter as it is not used for any of the checks below. So can't the `create_user_with_trial_instance()` method be a factory method that accepts a count and returns the `count` number of instances, which can be assigned like

```python
instance1, instance2, instance3 = self.user_trial_instance_factory(count=3)
```
?

Comment (spokerman12) [reply reports/tests.py]: True. Modified as requested.

**Review by lgp171188 — APPROVED** [review]
@spokerman12, :+1: Please wait till the unrelated CI failures are fixed before merging this PR.
- [x] I tested and verified that the periodic job now includes only the instances created in the last completed month.
- [x] I read through the code
- [x] ~~I checked for accessibility issues~~ NA
- [x] ~~Includes documentation~~ NA

**Comment by lgp171188** [comment]
@spokerman12, can you fix the failing quality check before moving the task to blocker?

**Comment by spokerman12** [comment]
@lgp171188 sure. I'd be in favor of adding a lint-fix for python; I don't believe it's implemented.

**Pull request merged**
